pdns/zonemd.cc

```diff
@@ -65,26 +28,117 @@ void pdns::zonemdVerify(const DNSName& zone, ZoneParserTNG& zpt, bool& validatio
       std::string err = "Bad record content in record for '" + dnsResourceRecord.qname.toStringNoDot() + "|" + dnsResourceRecord.qtype.toString() + "': " + e.what();
       throw PDNSException(err);
     }
-    if (dnsResourceRecord.qtype == QType::SOA && dnsResourceRecord.qname == zone) {
-      soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
+
+    if (dnsResourceRecord.qname == d_zone) {
+      switch (dnsResourceRecord.qtype) {
+      case QType::SOA:
+        d_soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
+        break;
+      case QType::DNSKEY:
+        d_dnskeys.emplace(std::dynamic_pointer_cast<DNSKEYRecordContent>(drc));
+        break;
+      case QType::ZONEMD: {
+        auto zonemd = std::dynamic_pointer_cast<ZONEMDRecordContent>(drc);
+        auto inserted = d_zonemdRecords.insert({pair(zonemd->d_scheme, zonemd->d_hashalgo), {zonemd, false}});
+        if (!inserted.second) {
+          // Mark as duplicate
+          inserted.first->second.duplicate = true;
+        }
+        break;
+      }
+      case QType::RRSIG: {
+        auto rrsig = std::dynamic_pointer_cast<RRSIGRecordContent>(drc);
+        if (rrsig->d_type == QType::NSEC) {
+          d_nsecs.signatures.emplace_back(rrsig);
+        }
+        else if (rrsig->d_type == QType::NSEC3) {
+          d_nsecs3.signatures.emplace_back(rrsig);
+        }
+        d_rrsigs.emplace_back(rrsig);
+        break;
+      }
+      case QType::NSEC:
+        d_nsecs.records.emplace(std::dynamic_pointer_cast<NSECRecordContent>(drc));
+        break;
+      case QType::NSEC3:
+        d_nsecs3.records.emplace(std::dynamic_pointer_cast<NSEC3RecordContent>(drc));
+        break;
+      }
     }
-    if (dnsResourceRecord.qtype == QType::ZONEMD && dnsResourceRecord.qname == zone) {
-      auto zonemd = std::dynamic_pointer_cast<ZONEMDRecordContent>(drc);
-      auto inserted = zonemdRecords.insert({pair(zonemd->d_scheme, zonemd->d_hashalgo), {zonemd, false}});
+    RRSetKey_t key = std::pair(dnsResourceRecord.qname, dnsResourceRecord.qtype);
+    d_resourceRecordSets[key].push_back(drc);
+    d_resourceRecordSetTTLs[key] = dnsResourceRecord.ttl;
+  }
+}
+
+void pdns::ZoneMD::readRecords(const vector<DNSRecord>& records)
+{
+  for (auto& record : records) {
+    readRecord(record);
+  }
+}
+
+void pdns::ZoneMD::readRecord(const DNSRecord& record)
+{
+  if (!record.d_name.isPartOf(d_zone) && record.d_name != d_zone) {
+    return;
+  }
+  if (record.d_type == QType::SOA && d_soaRecordContent) {
+    return;
+  }
+
+  if (record.d_name == d_zone) {
+    switch (record.d_type) {
+    case QType::SOA:
```

Comment:
I wonder if we could not merge that code with the one looking quite similar above for `void pdns::ZoneMD::readRecords(ZoneParserTNG& zpt)`?